smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,515 @@
+        if self._disposable:
+            return True
+        if self._batch_size <= 0:
+            return False
+        return self.qsize() >= self._batch_size
+
+    def empty(self) -> bool:
+        """Return True if the queue has 0 elements"""
+        return self.qsize() == 0
+
+
+class RequestDispatcher(Service):
+    def __init__(
+        self,
+        batch_timeout: float,
+        batch_size: int,
+        config_loader: EnvironmentConfigLoader,
+        worker_type: t.Type[MachineLearningWorkerBase],
+        comm_channel_type: t.Type[CommChannelBase] = DragonCommChannel,
+    ) -> None:
+        """The RequestDispatcher intercepts inference requests, stages them in
+        queues and batches them together before making them available to Worker
+        Managers.
+        :param batch_timeout: Maximum elapsed time before flushing a complete or incomplete batch
+        :param batch_size: Total capacity of each batch queue.
+        :param mem_pool: Memory pool used to share batched input tensors with worker
+        managers
+        :param config_loader: Object to load configuration from environment
+        :param worker_type: Type of worker to instantiate to batch inputs
+        :param comm_channel_type: Type of channel used to get requests
+        :raises SmartSimError: If config_loaded.get_queue() does not return a channel
+        """
+        super().__init__(as_service=True, cooldown=1)
+        self._queues: dict[str, list[BatchQueue]] = []
+        """Dict of all batch queues available for a given model key"""
+        self._active_queues: dict[str, BatchQueue] = {}
+        """Mapping telling which queue is the recipient of requests for a given model
+        key"""
+        self._batch_timeout = batch_timeout
+        """Time in seconds that has to be waited before flushing a non-full queue"""
+        self._batch_size = batch_size
+        """Total capacity of each batch queue."""
+        self._queue_swap_lock: t.Optional[RLock] = None
+        """Lock used to swap the active queue for a key"""
+        incoming_channel = config_loader.get_queue()
+        if incoming_channel is None:
+            raise SmartSimError("No incoming channel for dispatcher")
+        self._incoming_channel = incoming_channel
+        """The channel the dispatcher monitors for new tasks"""
+        self._outgoing_queue: DragonQueue = mp.Queue(maxsize=0)
+        """The queue on which batched inference requests are placed"""
+        self._feature_store: t.Optional[FeatureStore] = (
+            config_loader.get_feature_store()
+        )
+        """A feature store to retrieve models from"""
+        self._comm_channel_type = comm_channel_type
+        """The type of the channel used to receive requests"""
+        self._worker = worker_type()
+        """The worker used to batch inputs"""
+        self._mem_pool = MemoryPool.attach(dragon_gs_pool.create(2 * 1024**3).sdesc)
```

Comment:
How was this value chosen?  Should we define it as a constant somewhere?

Reply:
I put it as a parameter now, as I think it will be something to play with